statement.py

```diff
@@ -4,43 +4,272 @@
+                else:
+                    line.invoice = None
+            self.lines = lines
+
+    @classmethod
+    def validate_origin(cls, origins):
+        pool = Pool()
+        Line = pool.get('account.statement.line')
+        Warning = pool.get('res.user.warning')
+
+        paid_cancelled_invoice_lines = []
+        for origin in origins:
+            paid_cancelled_invoice_lines.extend(l for l in origin.lines
+                if l.invoice and l.invoice.state in {'cancelled', 'paid'})
+
+        if paid_cancelled_invoice_lines:
+            warning_key = Warning.format(
+                'statement_paid_cancelled_invoice_lines',
+                paid_cancelled_invoice_lines)
+            if Warning.check(warning_key):
+                raise StatementValidateWarning(warning_key,
+                    gettext('account_statement'
+                        '.msg_statement_invoice_paid_cancelled'))
+            Line.write(paid_cancelled_invoice_lines, {
+                    'related_to': None,
+                    })
+
+    def validate_amount(self):
+        pool = Pool()
+        Lang = pool.get('ir.lang')
+
+        amount = sum(l.amount for l in self.lines)
+        if amount != self.amount:
+            lang = Lang.get()
+            total_amount = lang.currency(
+                self.amount, self.statement.journal.currency)
+            amount = lang.currency(amount, self.statement.journal.currency)
+            raise StatementValidateError(
+                gettext('account_statement_enable_banking.'
+                    'msg_origin_pending_amount',
+                    origin_amount=total_amount,
+                    line_amount=amount))
+
+    @classmethod
+    def validate_statement(cls, statements):
+        pool = Pool()
+        Statement = pool.get('account.statement')
+        Line = pool.get('account.statement.line')
+        Warning = pool.get('res.user.warning')
+
+        paid_cancelled_invoice_lines = []
+        for statement in statements:
+            paid_cancelled_invoice_lines.extend(l for l in statement.lines
+                if l.invoice and l.invoice.state in {'cancelled', 'paid'})
+
+        if paid_cancelled_invoice_lines:
+            warning_key = Warning.format(
+                'statement_paid_cancelled_invoice_lines',
+                paid_cancelled_invoice_lines)
+            if Warning.check(warning_key):
+                raise StatementValidateWarning(warning_key,
+                    gettext('account_statement'
+                        '.msg_statement_invoice_paid_cancelled'))
+            Line.write(paid_cancelled_invoice_lines, {
+                    'related_to': None,
+                    })
+
+        Statement.write(statements, {
+                'state': 'validated',
+                })
+
     @classmethod
     @ModelView.button
     def reconcile(cls, origins):
         pool = Pool()
-        Line = pool.get('account.statement.line')
+        Statement = pool.get('account.statement')
```

Comment:
We need a comment here referencing that it is basically a copy & paste from `account_statement` `create_move()`.

Also I think we must rename this method `create_move()`.

And we need to add a state field to origin and call `create_move()` when moving to `posted` state.

Reply:
> Also I think we must rename this method `create_move()`.

The name of the function is the same as the button, if we change the function name, we have to change the button and the translation. For me is better the "reconcile" name.

Reply:
> And we need to add a state field to origin and call `create_move()` when moving to `posted` state.

As I say a few lines above, I wanted to avoid creating a state. But I agree that will be the better solution.